### PR TITLE
Update OpenId Connect Challenge Tests

### PR DESCRIPTION
**Pull request by troydai**
1. Expand the test coverage: add tests covers events work flow.
2. Move OpenID connect challenge tests to their own class.
3. Further refactory the test settings and utilities.

Issue: https://github.com/aspnet/Security/issues/772

Note:

1. The test utilities are reorganized for future reuse.
2. The leftover tests in `OpenIdMiddlewareTests` are not updated to keep the coverage for now. Those are signout tests. The test coverage for that flow will be updated later.

### CODE REVIEW

File: test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/ExpectedQueryValues.cs [hunk 1]
```diff
@@ -1,175 +0,0 @@
...
-using System.Text.Encodings.Web;
-using Microsoft.IdentityModel.Protocols.OpenIdConnect;
-using Xunit;
-
-namespace Microsoft.AspNetCore.Authentication.Tests.OpenIdConnect
-{
-    /// <summary>
-    /// This helper class is used to check that query string parameters are as expected.
-    /// </summary>
-    public class ExpectedQueryValues
```
Comment (troydai) [inline test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/ExpectedQueryValues.cs]: Renamed to `TestSettings`

File: test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/MockOpenIdConnectMessage.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+using Microsoft.IdentityModel.Protocols.OpenIdConnect;
+
+namespace Microsoft.AspNetCore.Authentication.Tests.OpenIdConnect
+{
+    internal class MockOpenIdConnectMessage : OpenIdConnectMessage
```
Comment (troydai) [inline test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/MockOpenIdConnectMessage.cs]: Extract from `OpenIdMiddlewareTests.cs`

File: test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/TestTransaction.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+namespace Microsoft.AspNetCore.Authentication.Tests.OpenIdConnect
+{
+    internal class TestTransaction
+    {
+        public static Task<TestTransaction> SendAsync(TestServer server, string url)
+        {
+            return SendAsync(server, url, cookieHeader: null);
+        }
+
+        public static async Task<TestTransaction> SendAsync(TestServer server, string uri, string cookieHeader)
```
Comment (HaoK) [inline test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/TestTransaction.cs]: Nit: you could make these extension methods on TestServer

File: test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/TestDefaultValues.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.IdentityModel.Protocols;
+using Microsoft.IdentityModel.Protocols.OpenIdConnect;
+
+namespace Microsoft.AspNetCore.Authentication.Tests.OpenIdConnect
+{
+    internal class TestDefaultValues
```
Comment (HaoK) [inline test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/TestDefaultValues.cs]: Nit: This class still seems like an oddball with all statics/constants, maybe it just can be merged with TestServerBuilder

File: test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/MockOpenIdConnectMessage.cs (hunk 2 above)
Comment (HaoK) [inline test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/MockOpenIdConnectMessage.cs]: Why do we need this, can't we just use `Mock<OpenIdConnectMessage>`?

File: test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/TestDefaultValues.cs (hunk 4 above)
Comment (troydai) [inline test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/TestDefaultValues.cs]: Yup, will do.

File: test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/MockOpenIdConnectMessage.cs (hunk 2 above)
Comment (troydai) [inline test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/MockOpenIdConnectMessage.cs]: Mock is supported under .net core? I'd love to use that then.

Comment (Tratcher) [inline test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/MockOpenIdConnectMessage.cs]: Please no.... we purged it from this repo

Comment (HaoK) [inline test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/MockOpenIdConnectMessage.cs]: Why is this repo anti Mock?  Just seems so much nicer than classes like this :/

Comment (Tratcher) [inline test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/MockOpenIdConnectMessage.cs]: Many of the usages of Mock that used to be here were 10x more code than alternatives. Most of the data structures we use are mutable and don't require mocking. You could use Mock, but with reasonable restraint please.

**Comment by troydai** [comment]
Also /cc @Tratcher

File: test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/OpenIdConnectChallengeTests.cs [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+        {
+            var settings = new TestSettings(
+                opts =>
+                {
+                    opts.Events = new OpenIdConnectEvents()
+                    {
+                        OnRedirectToIdentityProvider = context =>
+                        {
+                            context.HandleResponse();
+                            return Task.FromResult(0);
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/OpenIdConnectChallengeTests.cs]: Set a custom status code and header.

File: test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/OpenIdConnectChallengeTests.cs [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+        public async Task OnRedirectToIdentityProviderEventSkipResponse()
+        {
+            var settings = new TestSettings(
+                opts =>
+                {
+                    opts.Events = new OpenIdConnectEvents()
+                    {
+                        OnRedirectToIdentityProvider = context =>
+                        {
+                            context.SkipToNextMiddleware();
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/OpenIdConnectChallengeTests.cs]: Is this functionally different from HandleResponse in this context?

File: test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/OpenIdConnectChallengeTests.cs [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+            var firstCookie = transaction.SetCookie.First();
+            Assert.Contains(OpenIdConnectDefaults.CookieNoncePrefix, firstCookie);
+            Assert.Contains("expires", firstCookie);
+
+            var secondCookie = transaction.SetCookie.Skip(1).First();
+            Assert.StartsWith(".AspNetCore.Correlation.OpenIdConnect.", secondCookie);
+            Assert.Contains("expires", secondCookie);
+        }
+
+        private static string ChallengeEndpoint => TestServerBuilder.TestHost + TestServerBuilder.Challenge;
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/OpenIdConnectChallengeTests.cs]: Move to the top of the fil

**Comment by Tratcher** [comment]
:shipit:

File: test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/OpenIdConnectChallengeTests.cs (hunk 6 above)
Comment (troydai) [inline test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/OpenIdConnectChallengeTests.cs]: See the above comments. For the test as it is, they're the same. But I'd like to distinguish them for future expansion of this test.

File: test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/OpenIdConnectChallengeTests.cs (hunk 5 above)
Comment (troydai) [inline test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/OpenIdConnectChallengeTests.cs]: ok.

**Pull request merged**